main.go

```diff
@@ -134,21 +154,77 @@ func main() {
 			},
 		}
 	} else {
-		log.Println("Using parameters from the configuration file")
+		log.Info().Msg("Using parameters from the configuration file")
 		if *args.pattern != "" || *args.file != "" {
-			log.Fatalln("No pattern/file allowed when configuration file is provided, aborting")
+			log.Fatal().Msg("No pattern/file allowed when configuration file is provided, aborting")
 		}
 
 		configPath := *args.config
 		parsedConfig, err := ReadConfigFromFile(configPath)
 		if err != nil {
-			log.Fatal(err)
+			log.Fatal().Err(err).Msg("Failed reading configuration from file")
 		}
 		config = parsedConfig
-		log.Printf("Configuration file loaded: \"%s\"\n", configPath)
+		log.Info().Msgf("Configuration file loaded: \"%s\"\n", configPath)
 	}
 
 	initSentry(config)
-	catchInterrupt()
-	runWithConfig(config)
+
+	// Create global condition
+	cond := sync.NewCond(&sync.Mutex{})
+
+	// Listen to interrupt/termination signal.
+	// Once received, this will exit the program.
+	exitSignal := make(chan os.Signal, 1)
+	signal.Notify(exitSignal, syscall.SIGTERM, os.Interrupt, os.Kill)
+
+	// Start the main program
+	log.Debug().Msg("Starting sentlog")
+	go runWithConfig(config, cond)
+
+	// Listen to SIGHUP for handling graceful restarts.
+	// This is most useful on avoiding logrotate not be able to rotate the log file.
+	// It's an unbuffered channel, you can send SIGHUP multiple times.
+	usrStopSignal := make(chan os.Signal)
+	signal.Notify(usrStopSignal, syscall.SIGHUP)
+	go func() {
+		for {
+			log.Debug().Msg("Waiting for SIGHUP")
+			<-usrStopSignal
+
+			log.Debug().Msg("Received SIGHUP")
+			cond.L.Lock()
+			_killed = true
+			cond.Broadcast()
+			cond.L.Unlock()
+
+			log.Debug().Msg("Sleeping for 10 seconds")
+			// NOTE: should this be configurable in the future?
+			time.Sleep(time.Second * 10)
```

Comment:
I really need feedback about this. I actually wanted to use `SIGUSR1` and `SIGUSR2`. But realizing that it's not available on Windows (and I'm developing this on Windows too), I chose `SIGHUP` instead.

If I were to use `SIGUSR1` and `SIGUSR2`, we might use `SIGUSR1` to stop sentlog and `SIGUSR2` to start it again. But considering this, people might use Windows and tail their Windows application logs or something.

I think this really needs to be configured, with default time of 10 seconds.